relay-cardinality/src/cache.rs

```diff
@@ -0,0 +1,251 @@
+use std::hash::Hash;
+use std::sync::{PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use crate::OrganizationId;
+
+/// Cached outcome, wether the item can be accepted, rejected or the cache has no information about
+/// this hash.
+#[derive(Debug)]
+pub enum CacheOutcome {
+    /// Hash accepted by cache.
+    Accepted,
+    /// Hash rejected by cache.
+    Rejected,
+    /// Cache has no information about the hash.
+    Unknown,
+}
+
+/// Internal cache remembering already accepted elements and current cardinality.
+///
+/// Only caches for the currently active granule of the sliding window.
+#[derive(Default)]
+pub struct Cache {
+    inner: RwLock<Inner>,
+}
+
+impl Cache {
+    /// Acquires a read lock from the cache and returns a read handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheRead`] must be dropped.
+    pub fn read(
+        &self,
+        organization_id: OrganizationId,
+        window: u64,
+        limit: usize,
+    ) -> CacheRead<'_> {
+        let inner = self.inner.read().unwrap_or_else(PoisonError::into_inner);
+
+        // If the window does not match, we can already release the lock.
+        if window != inner.current_window {
+            return CacheRead::noop();
+        }
+
+        CacheRead::new(inner, organization_id, limit)
+    }
+
+    /// Acquires a write lock from the cache and returns an update handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheUpdate`] must be dropped.
+    pub fn update<'a>(
+        &'a self,
+        organization_id: OrganizationId,
+        item_scope: &'a str,
+        window: u64,
+    ) -> CacheUpdate<'a> {
+        let mut inner = self.inner.write().unwrap_or_else(PoisonError::into_inner);
+
+        // If the window is older don't do anything and give up the lock early.
+        if window < inner.current_window {
+            return CacheUpdate::noop();
+        }
+
+        // If the window is newer then the current window, reset the cache to the new window.
+        if window > inner.current_window {
+            inner.current_window = window;
+            inner.cache.clear();
+        }
+
+        let key = CacheKeyRef {
+            organization_id,
+            item_scope,
+        };
+        CacheUpdate::new(inner, key)
+    }
+}
+
+/// Cache read handle.
+///
+/// Holds a cache read lock, the lock is released on drop.
+pub struct CacheRead<'a>(CacheReadInner<'a>);
```

Comment:
Wrapper struct is new, `CacheReadInner` can't be public without making `Inner` public, which is not what we want.